enspara/geometry/rotamer.py

```diff
@@ -31,44 +27,198 @@ def dihedral_angles(traj, dihedral_type):
+        a single dihedral angle from a trajectory. This array MUST: 
+        a) Be passed in degrees 
+        b) Span from 0 to 360 in range 
+    hard_boundaries : array-like, shape=(n_boundaries)
+        Single set of numbers containing the "hard" boundaries denoting 
+        rotamer basins. This array MUST contain 0 as the first value and 
+        360 as the last value. 
+    buffer_width : int, default=15
+        size of the buffer region on either side of the 
+        rotamer barrier. 
+
+    Returns
+    --------
+    rotamers : array-like, shape=(n_frames) 
+        Time-series data with rotamer state assignments. Each element i 
+        contains the rotamer state assignment computed using the "angles" 
+        array. 
+
+    See also
+    ---------
+    is_buffered_transition, get_gates
+    """
     n_basins = len(hard_boundaries)-1
 
     if buffer_width <= 0 or buffer_width >= 360. / n_basins:
-        return None
+        raise DataInvalid('Buffer width must be between 0 and 360 degrees.')
+        #return None
     if hard_boundaries[0] != 0 or hard_boundaries[-1] != 360:
-        return None
-
-    core_edges = [0]
-    n_cutoffs = len(hard_boundaries)
-    for i in range(n_cutoffs-1):
-        core_edges.append(hard_boundaries[i]+buffer_width)
-        core_edges.append(hard_boundaries[i+1]-buffer_width)
-    core_edges.append(360)
-
-    bins = np.digitize(angles, core_edges)
+        raise DataInvalid('hard_boundaries list must start with 0 and end with 360')
+        #return None
 
+    # Need to establish how long it is and create an appropriate output array
     n_frames = len(angles)
     rotamers = -1*np.ones(n_frames, dtype='int16')
 
-    # assign things to cores
-    for i in range(n_basins):
-        inds = np.where(bins == 2*(i+1))[0]
-        rotamers[inds] = i
-
+    # First, assign the first state to its rotamer bin
     # make sure assign first frame
     for i in range(n_basins):
         if angles[0] < hard_boundaries[i+1]:
             rotamers[0] = i
             break
 
-    # do one sweep to make sure verythign is assigned to a bin
+    # Now we will go through each subsequent element of the array and 
+    # assign each state based on whether or not there is a buffered transition
+    cur_state = rotamers[0]
+    #transition_Array= [] used for debugging transitions in rotamers
     for i in range(1, n_frames):
-        if rotamers[i] == -1:
-            rotamers[i] = rotamers[i-1]
+        new_angle = angles[i]
+        cur_angle = angles[i-1]
+
+        # If there is a buffered transition we will reassign states
+        if is_buffered_transition(cur_state, new_angle, hard_boundaries, buffer_width):
+            cur_state = np.digitize(new_angle, hard_boundaries) - 1
+
+        rotamers[i] = cur_state
 
     return rotamers
 
+def is_buffered_transition(cur_state, new_angle, hard_boundaries, buffer_width):
+    """Returns whether or not the change in angle is a buffered transition
+
```

Comment:
Can we add an additional sentence or two as to what "buffered" transition means?